pkg/provider/infisical/provider.go

```diff
@@ -0,0 +1,144 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implieclient.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package infisical
+
+import (
+	"context"
+	"errors"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	esmeta "github.com/external-secrets/external-secrets/apis/meta/v1"
+	"github.com/external-secrets/external-secrets/pkg/provider/infisical/api"
+	"github.com/external-secrets/external-secrets/pkg/provider/infisical/constants"
+	"github.com/external-secrets/external-secrets/pkg/utils"
+	"github.com/external-secrets/external-secrets/pkg/utils/resolvers"
+)
+
+var (
+	Logger = ctrl.Log.WithName("provider").WithName(constants.ProviderName)
+)
+
+type Provider struct {
+	apiClient api.InfisicalApis
+	apiScope  *InfisicalClientScope
+}
+
+type InfisicalClientScope struct {
+	SecretPath      string
+	ProjectSlug     string
+	EnvironmentSlug string
+}
+
+// https://github.com/external-secrets/external-secrets/issues/644
+var _ esv1beta1.SecretsClient = &Provider{}
+var _ esv1beta1.Provider = &Provider{}
+
+func init() {
+	esv1beta1.Register(&Provider{}, &esv1beta1.SecretStoreProvider{
+		Infisical: &esv1beta1.InfisicalProvider{},
+	})
+}
+
+func (p *Provider) Capabilities() esv1beta1.SecretStoreCapabilities {
+	return esv1beta1.SecretStoreReadOnly
+}
+
+func (p *Provider) NewClient(ctx context.Context, store esv1beta1.GenericStore, kube kclient.Client, namespace string) (esv1beta1.SecretsClient, error) {
+	storeSpec := store.GetSpec()
+
+	if storeSpec == nil || storeSpec.Provider == nil || storeSpec.Provider.Infisical == nil {
+		return nil, errors.New("invalid infisical store")
+	}
+
+	infisicalSpec := storeSpec.Provider.Infisical
+
+	apiClient, err := api.NewAPIClient(infisicalSpec.HostAPI)
+	if err != nil {
+		return nil, err
+	}
+
+	if infisicalSpec.Auth.Type == constants.UniversalAuth {
+		universalAuthCredentials := infisicalSpec.Auth.UniversalAuthCredentials
```

Comment:
we need to check if `UniversalAuthCredentials` is not `nil`, as users can set up the correct Type but not pass any credentials. This will cause (if I understood it correctly) a panic.

Which is why I personally think it is bette rto not have an `auth.Type` in the first place, and just implicit order (e.g. -> JWT/OIDC, -> Workload Identity -> UniversalAuth)

Reply:
Ohh ok. I think this ans above question i have raised.